### ISSUE TITLE
L001: Using jinja & dbt_utils results in unwanted unneccessary trailing whitespace with sqlfluff 1.3.2

### ISSUE THREAD

**Issue by FrankTub** [sqlfluff/sqlfluff#3988]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I'm trying to move to latest dbt & sqlfluff version. I now use `dbt 1.2.1`  & `sqlfluff 1.2.1 `, and trying to move to `dbt 1.3.0` & `sqlfluff 1.3.2` . However when using the latest versions I keep running into a lot of issues like I mention in `Observed Behavior` that were not present in the versions I currently use.

### Expected Behaviour

The expected behavior would be to have no linting error in this scenario like I used to have in version `sqlfluff 1.2.1`.

### Observed Behaviour

My `dim_dates.sql` file that I'm linting:

```sql
{{ config(materialized="table") }}

with
date_spine as (
    select date_actual
    from {{ ref('dates') }}
),

calculated as (
    select
        date_actual,
        extract(isodow from date_actual) as day_of_week,
        extract(day from date_actual) as day_of_month,
        extract(doy from date_actual) as day_of_year,
        to_char(date_actual, 'TMDay') as day_name,
        extract(month from date_actual) as month_actual,
        to_char(date_actual, 'TMMonth') as month_name,
        extract(year from date_actual) as year_actual,
        extract(week from date_actual) as week_of_year,
        'Q' || to_char(date_actual, 'Q') as quarter_actual,
        to_char(date_actual, 'iyyy/IW') as year_week,
        to_char(date_actual, 'yyyy/mm') as year_month,
        to_char(date_actual, 'yyyy/"Q"Q') as year_quarter,
        case when extract(isodow from date_actual) in (6, 7) then 0 else 1 end as is_weekday,
        case when extract(isodow from date_actual) in (6, 7) then 1 else 0 end as is_weekend,
        {{ dbt_utils.current_timestamp() }} as dbt_created_at
    from date_spine
)

select *
from calculated
```

The result of linting:

```bash
== [models/datamarts/dim_dates/dim_dates.sql] FAIL                                                                                                                                                                       
L:  26 | P:   1 | L001 | Unnecessary trailing whitespace.
```

It is complaining about the following. It does not allow a space on the line with column `dbt_created_at`. It now is:

```sql
        {{ dbt_utils.current_timestamp() }} as dbt_created_at
```

When running `sqlfluff fix` it turns into:

```sql
{{ dbt_utils.current_timestamp() }} as dbt_created_at
```

### How to reproduce

You need two sql files & a dbt package.

`dates.sql`:

```sql
{{ config(materialized="table") }}

with
date_spine as (
  {{ dbt_utils.date_spine(
      start_date="to_date('01/01/2010', 'mm/dd/yyyy')",
      datepart="day",
      end_date="CURRENT_DATE + INTERVAL '12 years'"
     )
  }}
),

all_dates as (
    select
        date_day as date_actual,
        {{ dbt_utils.current_timestamp() }} as dbt_created_at
    from date_spine
)

select *
from all_dates

```

`dim_dates.sql`:

```sql
{{ config(materialized="table") }}

with
date_spine as (
    select date_actual
    from {{ ref('dates') }}
),

calculated as (
    select
        date_actual,
        extract(isodow from date_actual) as day_of_week,
        extract(day from date_actual) as day_of_month,
        extract(doy from date_actual) as day_of_year,
        to_char(date_actual, 'TMDay') as day_name,
        extract(month from date_actual) as month_actual,
        to_char(date_actual, 'TMMonth') as month_name,
        extract(year from date_actual) as year_actual,
        extract(week from date_actual) as week_of_year,
        'Q' || to_char(date_actual, 'Q') as quarter_actual,
        to_char(date_actual, 'iyyy/IW') as year_week,
        to_char(date_actual, 'yyyy/mm') as year_month,
        to_char(date_actual, 'yyyy/"Q"Q') as year_quarter,
        case when extract(isodow from date_actual) in (6, 7) then 0 else 1 end as is_weekday,
        case when extract(isodow from date_actual) in (6, 7) then 1 else 0 end as is_weekend,
        {{ dbt_utils.current_timestamp() }} as dbt_created_at
    from date_spine
)

select *
from calculated
```

`packages.yml`:

```yml
packages:
  - package: dbt-labs/dbt_utils
    version: 0.8.6
```

My guess would be that this is irrespective of sqlfluff configuration, but will deliver my `.sqlfluff` below.

### Dialect

postgres

### Version

Python version: `3.10.6`
sqlfluff version: `1.3.2`
dbt version: `1.3.0`

### Configuration

`.sqlfluff`:

```
[sqlfluff]
# verbose is an integer (0-2) indicating the level of log output
verbose = 0
# Turn off color formatting of output
nocolor = False
# Supported dialects https://docs.sqlfluff.com/en/stable/dialects.html
# Or run 'sqlfluff dialects'
dialect = postgres
# One of [raw|jinja|python|placeholder]
templater = dbt
# Comma separated list of rules to check, default to all
rules = all
# Comma separated list of rules to exclude, or None
exclude_rules = L031
# The depth to recursively parse to (0 for unlimited)
recurse = 0
# Below controls SQLFluff output, see max_line_length for SQL output
output_line_length = 160
# Number of passes to run before admitting defeat
runaway_limit = 10
# Ignore errors by category (one or more of the following, separated by commas: lexing,linting,parsing,templating)
ignore = None
# Ignore linting errors found within sections of code coming directly from
# templated code (e.g. from within Jinja curly braces. Note that it does not
# ignore errors from literal code found within template loops.
ignore_templated_areas = True
# can either be autodetect or a valid encoding e.g. utf-8, utf-8-sig
encoding = autodetect
# Ignore inline overrides (e.g. to test if still required)
disable_noqa = False
# Comma separated list of file extensions to lint
# NB: This config will only apply in the root folder
sql_file_exts = .sql,.sql.j2,.dml,.ddl
# Allow fix to run on files, even if they contain parsing errors
# Note altering this is NOT RECOMMENDED as can corrupt SQL
fix_even_unparsable = False

[sqlfluff:indentation]
# See https://docs.sqlfluff.com/en/stable/indentation.html
indented_joins = False
indented_ctes = False
indented_using_on = True
template_blocks_indent = True

[sqlfluff:templater:dbt]
project_dir = .
profiles_dir = ~/.dbt

[sqlfluff:templater:jinja]
apply_dbt_builtins = True

# Some rules can be configured directly from the config common to other rules
[sqlfluff:rules]
tab_space_size = 4
max_line_length = 160
indent_unit = space
comma_style = trailing
allow_scalar = True
single_table_references = consistent
unquoted_identifiers_policy = all

# Some rules have their own specific config
[sqlfluff:rules:L003]
hanging_indents = True

[sqlfluff:rules:L007]
operator_new_lines = after

[sqlfluff:rules:L010]
# Keywords
capitalisation_policy = lower
# Comma separated list of words to ignore for this rule
ignore_words = None
ignore_words_regex = None

[sqlfluff:rules:L011]
# Aliasing preference for tables
aliasing = explicit

[sqlfluff:rules:L012]
# Aliasing preference for columns
aliasing = explicit

[sqlfluff:rules:L014]
# Unquoted identifiers
extended_capitalisation_policy = consistent
# Comma separated list of words to ignore for this rule
ignore_words = None
ignore_words_regex = None

[sqlfluff:rules:L016]
# Line length
ignore_comment_lines = True
ignore_comment_clauses = True

[sqlfluff:rules:L026]
# References must be in FROM clause
# Disabled for some dialects (e.g. bigquery)
force_enable = False

[sqlfluff:rules:L028]
# References must be consistently used
# Disabled for some dialects (e.g. bigquery)
force_enable = True
single_table_references = unqualified

[sqlfluff:rules:L029]
# Keywords should not be used as identifiers.
unquoted_identifiers_policy = aliases
quoted_identifiers_policy = none
# Comma separated list of words to ignore for this rule
ignore_words = date,year_month,name
ignore_words_regex = None

[sqlfluff:rules:L030]
# Function names
extended_capitalisation_policy = lower
# Comma separated list of words to ignore for this rule
ignore_words = None
ignore_words_regex = None

[sqlfluff:rules:L031]
# Avoid table aliases in from clauses and join conditions.
# Disabled for some dialects (e.g. bigquery)
force_enable = False

[sqlfluff:rules:L038]
# Trailing commas
select_clause_trailing_comma = forbid

[sqlfluff:rules:L040]
# Null & Boolean Literals
capitalisation_policy = lower
# Comma separated list of words to ignore for this rule
ignore_words = None
ignore_words_regex = None

[sqlfluff:rules:L042]
# By default, allow subqueries in from clauses, but not join clauses
forbid_subquery_in = both

[sqlfluff:rules:L047]
# Consistent syntax to count all rows
prefer_count_1 = False
prefer_count_0 = False

[sqlfluff:rules:L051]
# Fully qualify JOIN clause
fully_qualify_join_types = inner

[sqlfluff:rules:L052]
# Semi-colon formatting approach
multiline_newline = False
require_final_semicolon = False

[sqlfluff:rules:L054]
# GROUP BY/ORDER BY column references
group_by_and_order_by_style = explicit

[sqlfluff:rules:L057]
# Special characters in identifiers
unquoted_identifiers_policy = all
quoted_identifiers_policy = all
allow_space_in_identifier = False
additional_allowed_characters = ""
ignore_words = None
ignore_words_regex = None

[sqlfluff:rules:L059]
# Policy on quoted and unquoted identifiers
prefer_quoted_identifiers = False
ignore_words = None
ignore_words_regex = None
force_enable = False

[sqlfluff:rules:L062]
# Comma separated list of blocked words that should not be used
blocked_words = None

[sqlfluff:rules:L063]
# Data Types
extended_capitalisation_policy = consistent
# Comma separated list of words to ignore for this rule
ignore_words = None
ignore_words_regex = None

[sqlfluff:rules:L064]
# Consistent usage of preferred quotes for quoted literals
preferred_quoted_literal_style = consistent
# Disabled for dialects that do not support single and double quotes for quoted literals (e.g. Postgres)
force_enable = False
```

`.sqlfluffignore`:

```
target/
dbt_packages/
macros/
.cache/
venv/
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by pquadri** [comment]
similar issue,  i get the same error whenever a line starts with a macro and i put any indentation before the macro call

**Comment by alanmcruickshank** [comment]
@pquadri @FrankTub - I'm having trouble replicating this issue. Can you confirm whether it's possible to recreate the same issue using a single file implementation - or whether it only occurs in the multi file scenario? I've been testing with this file which raises no errors:

```sql
{% macro foo() -%}
    1 + 2
{%- endmacro %}

SELECT
    {{  foo() }} as baz
FROM bar
```

I suspect that #3942 may solve this issue, but I would love to confirm that by demonstrating it's solved.

**Comment by FrankTub** [comment]
@alanmcruickshank , I just simplified my query and can reproduce the issue in `sqlfluff 1.4.1` with following query.

```sql
{%- set json_keys = ["approve_sign", "complete_comply", "complete_dossier", "comply", "early_buyout", "payout", "prepare_payout", "sign"] -%}

with
dummy as (
    select
        {% for json_key in json_keys -%}
        case
            when 1 = 1
                {% if json_key in ["payout"] %}
                then 0
            {% else %}
                then 1
            {% endif %}
            else null
        end as {{ json_key }}_working_minutes{% if not loop.last %},{% endif %}
        {% endfor %}
)

select *
from dummy

```

**Comment by alanmcruickshank** [comment]
So I've linted that query on both the current main and on #3942.

### Current Main Branch

Linting output
```
== [test.sql] FAIL
L:  15 | P:  68 | L008 | Commas should be followed by a single whitespace unless
                       | followed by a comment.
L:  15 | P:  68 | L008 | Commas should be followed by a single whitespace unless
                       | followed by a comment.
L:  15 | P:  68 | L008 | Commas should be followed by a single whitespace unless
                       | followed by a comment.
L:  15 | P:  68 | L008 | Commas should be followed by a single whitespace unless
                       | followed by a comment.
L:  15 | P:  68 | L008 | Commas should be followed by a single whitespace unless
                       | followed by a comment.
L:  15 | P:  68 | L008 | Commas should be followed by a single whitespace unless
                       | followed by a comment.
L:  15 | P:  68 | L008 | Commas should be followed by a single whitespace unless
                       | followed by a comment.
L:  16 | P:   9 | L003 | Expected 1 indentation, found 2 [compared to line 03]
L:  20 | P:  11 | L009 | Files must end with a single trailing newline.
```

and the fixed query is:

```sql
{%- set json_keys = ["approve_sign", "complete_comply", "complete_dossier", "comply", "early_buyout", "payout", "prepare_payout", "sign"] -%}

with
dummy as (
    select
        {% for json_key in json_keys -%}
        case
            when 1 = 1
                {% if json_key in ["payout"] %}
                then 0
            {% else %}
                then 1
            {% endif %}
            else null
        end as {{ json_key }}_working_minutes{% if not loop.last %}, {% endif %}
    {% endfor %}
)

select *
from dummy
```

The issues here are that I can see are:
- The jinja tags aren't aligned.
- The whitespace-after-comma rule reports multiple times.

### On #3942 

Linting Output:
```
== [test.sql] FAIL
L:   6 | P:   1 | L003 | Expected indent of 4 spaces.
L:   7 | P:   9 | L003 | Expected indent of 4 spaces.
L:   7 | P:   9 | L003 | Expected indent of 4 spaces.
L:   7 | P:   9 | L003 | Expected indent of 4 spaces.
L:   7 | P:   9 | L003 | Expected indent of 4 spaces.
L:   7 | P:   9 | L003 | Expected indent of 4 spaces.
L:   7 | P:   9 | L003 | Expected indent of 4 spaces.
L:   7 | P:   9 | L003 | Expected indent of 4 spaces.
L:   7 | P:   9 | L003 | Expected indent of 4 spaces.
L:   8 | P:   1 | L003 | Expected indent of 8 spaces.
L:   8 | P:   1 | L003 | Expected indent of 8 spaces.
L:   8 | P:   1 | L003 | Expected indent of 8 spaces.
L:   8 | P:   1 | L003 | Expected indent of 8 spaces.
L:   8 | P:   1 | L003 | Expected indent of 8 spaces.
L:   8 | P:   1 | L003 | Expected indent of 8 spaces.
L:   8 | P:   1 | L003 | Expected indent of 8 spaces.
L:   8 | P:   1 | L003 | Expected indent of 8 spaces.
L:   8 | P:  17 | L003 | Expected line break and indent of 16 spaces before '1'.
L:   8 | P:  17 | L003 | Expected line break and indent of 16 spaces before '1'.
L:   8 | P:  17 | L003 | Expected line break and indent of 16 spaces before '1'.
L:   8 | P:  17 | L003 | Expected line break and indent of 16 spaces before '1'.
L:   8 | P:  17 | L003 | Expected line break and indent of 16 spaces before '1'.
L:   8 | P:  17 | L003 | Expected line break and indent of 16 spaces before '1'.
L:   8 | P:  17 | L003 | Expected line break and indent of 16 spaces before '1'.
L:   8 | P:  17 | L003 | Expected line break and indent of 16 spaces before '1'.
L:   9 | P:   1 | L003 | Expected indent of 8 spaces.
L:   9 | P:   1 | L003 | Expected indent of 8 spaces.
L:   9 | P:   1 | L003 | Expected indent of 8 spaces.
L:   9 | P:   1 | L003 | Expected indent of 8 spaces.
L:   9 | P:   1 | L003 | Expected indent of 8 spaces.
L:   9 | P:   1 | L003 | Expected indent of 8 spaces.
L:   9 | P:   1 | L003 | Expected indent of 8 spaces.
L:   9 | P:   1 | L003 | Expected indent of 8 spaces.
L:  11 | P:   1 | L003 | Expected indent of 8 spaces.
L:  13 | P:   1 | L003 | Expected indent of 8 spaces.
L:  13 | P:   1 | L003 | Expected indent of 8 spaces.
L:  13 | P:   1 | L003 | Expected indent of 8 spaces.
L:  13 | P:   1 | L003 | Expected indent of 8 spaces.
L:  13 | P:   1 | L003 | Expected indent of 8 spaces.
L:  13 | P:   1 | L003 | Expected indent of 8 spaces.
L:  13 | P:   1 | L003 | Expected indent of 8 spaces.
L:  14 | P:   1 | L003 | Expected indent of 8 spaces.
L:  14 | P:   1 | L003 | Expected indent of 8 spaces.
L:  14 | P:   1 | L003 | Expected indent of 8 spaces.
L:  14 | P:   1 | L003 | Expected indent of 8 spaces.
L:  14 | P:   1 | L003 | Expected indent of 8 spaces.
L:  14 | P:   1 | L003 | Expected indent of 8 spaces.
L:  14 | P:   1 | L003 | Expected indent of 8 spaces.
L:  14 | P:   1 | L003 | Expected indent of 8 spaces.
L:  15 | P:   1 | L003 | Expected indent of 4 spaces.
L:  15 | P:   1 | L003 | Expected indent of 4 spaces.
L:  15 | P:   1 | L003 | Expected indent of 4 spaces.
L:  15 | P:   1 | L003 | Expected indent of 4 spaces.
L:  15 | P:   1 | L003 | Expected indent of 4 spaces.
L:  15 | P:   1 | L003 | Expected indent of 4 spaces.
L:  15 | P:   1 | L003 | Expected indent of 4 spaces.
L:  15 | P:   1 | L003 | Expected indent of 4 spaces.
L:  15 | P:  46 | L003 | Expected line break and indent of 4 spaces before ''.
L:  15 | P:  68 | L008 | Commas should be followed by a single whitespace unless
                       | followed by a comment.
L:  15 | P:  68 | L008 | Commas should be followed by a single whitespace unless
                       | followed by a comment.
L:  15 | P:  68 | L008 | Commas should be followed by a single whitespace unless
                       | followed by a comment.
L:  15 | P:  68 | L008 | Commas should be followed by a single whitespace unless
                       | followed by a comment.
L:  15 | P:  68 | L008 | Commas should be followed by a single whitespace unless
                       | followed by a comment.
L:  15 | P:  68 | L008 | Commas should be followed by a single whitespace unless
                       | followed by a comment.
L:  15 | P:  68 | L008 | Commas should be followed by a single whitespace unless
                       | followed by a comment.
L:  16 | P:   1 | L003 | Expected indent of 4 spaces.
L:  16 | P:   1 | L003 | Expected indent of 4 spaces.
L:  16 | P:   1 | L003 | Expected indent of 4 spaces.
L:  16 | P:   1 | L003 | Expected indent of 4 spaces.
L:  16 | P:   1 | L003 | Expected indent of 4 spaces.
L:  16 | P:   1 | L003 | Expected indent of 4 spaces.
L:  16 | P:   1 | L003 | Expected indent of 4 spaces.
L:  16 | P:   1 | L003 | Expected indent of 4 spaces.
L:  20 | P:  11 | L009 | Files must end with a single trailing newline.
```

and the fixed query is:

```sql
{%- set json_keys = ["approve_sign", "complete_comply", "complete_dossier", "comply", "early_buyout", "payout", "prepare_payout", "sign"] -%}

with
dummy as (
    select
    {% for json_key in json_keys -%}
    case
        when
            1 = 1
        {% if json_key in ["payout"] %}
            then 0
        {% else %}
            then 1
        {% endif %}
        else null
    end as {{ json_key }}_working_minutes{% if not loop.last %}, {% endif %}
    {% endfor %}
)

select *
from dummy
```

The issues I see here:
- We're still getting multiple reports of the same issue (I'll make a separate issue for that).
- While the Jinja tags _are_ aligned, the whole query gets stepped back one step - which is strange.

@FrankTub - do I interpret that your main issue is that the jinja tags aren't aligned? If so I think that's fixed in the PR branch, but there is still the issue of how this file should mostly be indented.

### What should _actually_ happen:

I assume the _correct_ layout of this file would be:

```sql
{%- set json_keys = ["approve_sign", "complete_comply", "complete_dossier", "comply", "early_buyout", "payout", "prepare_payout", "sign"] -%}

with
dummy as (
    select
        {% for json_key in json_keys -%}
            case
                when 1 = 1
                    {% if json_key in ["payout"] %}
                        then 0
                    {% else %}
                        then 1
                    {% endif %}
                else null
            end as {{ json_key }}_working_minutes{% if not loop.last %}, {% endif %}
        {% endfor %}
)

select *
from dummy
```

Is that in line with your expectations?

**Comment by alanmcruickshank** [comment]
Assuming it is, I've added this as a test case in #3942 which should now resolve this.

**Comment by FrankTub** [comment]
@alanmcruickshank , my expected output would be something like.

```sql
{%- set json_keys = ["approve_sign", "complete_comply", "complete_dossier", "comply", "early_buyout", "payout", "prepare_payout", "sign"] -%}

with
dummy as (
    select
        {% for json_key in json_keys -%}
        case
            when 1 = 1
            {% if json_key in ["payout"] %}
                then 0
            {% else %}
                then 1
            {% endif %}
            else null
        end as {{ json_key }}_working_minutes{% if not loop.last %},{% endif %}
        {% endfor %}
)

select *
from dummy
```

And I would prefer it that way so that my sql statements are with indentations as I would expect, meaning it will be fixed to something like:

```sql
with
dummy as (
    select
        case
            when 1 = 1
                then 1
            else null
        end as approve_sign_working_minutes,
...
        case
            when 1 = 1
                then 1
            else null
        end as sign_working_minutes

)

select *
from dummy
```

This differs compared to your formatting, since there you have an indentation for a jinja statement, as well as for the sql code in the next line. That would result in compiled SQL like:

```sql
with
dummy as (
    select
            case
                when 1 = 1
                        then 1
                else null
        end as approve_sign_working_minutes,
...
            case
                when 1 = 1
                        then 1
                else null
        end as approve_sign_working_minutes,

)

select *
from dummy
```

This is a not a match with how I would like my compiled SQL to look like.

**Comment by alanmcruickshank** [comment]
Interesting. So that I've got some more context - what's your use case where the layout of the rendered SQL is also important? That's not a common question we get.

**Comment by FrankTub** [comment]
Fair point, it's not a requirement to have the compiled SQL with correct indentation, it's just my preference when trying to write my code based on a compiled model that the formatting is OK.

That having said, I would really like sqlfluff to auto-fix jinja sql with the indentations as you have proposed! It makes the dbt model code way more readable.

**Comment by alanmcruickshank** [comment]
Ok - understood. The challenge here is that making both the _rendered_ and _raw_ sql both be indented sensibly ends up being a conflicting aim. We've prioritised for the _raw_ sql to make sense, because that's the main focus of the tool. You _could_ use the `template_blocks_indent` configuration value (see [docs](https://docs.sqlfluff.com/en/stable/configuration.html)), so make both _raw_ and _rendered_ closer, but I don't think it will quite get you what you want.

The way I've suggested above, is coming in 2.0.0 (it's not there yet, the existing indentation algorithm has needed an overhaul for a while).

**Comment by FrankTub** [comment]
Yeah I think it is definitely an improvement, I agree that the raw sql is by far the most important to be formatted in a sensible way. Keep up the good work!
